<div style="border:solid Chocolate 2px; padding: 40px">
 
<b> Иван, привет!👋</b> 

Меня зовут Кирилл Васильев, я буду ревьюером твоего проекта. Я предлагаю общаться на «ты», но если привычнее на «вы», просто скажи об этом! 

Я буду оставлять комментарии в твоем проекте. Пожалуйста, не удаляй их, они могут быть полезны в случае повторной проверки.

Свои комментарии я буду обозначать <font color='green'>зеленым</font>, <font color='gold'>желтым</font> и <font color='red'>красным</font> цветами, например:
  
<br/> 

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> В случае, если решение на отдельном шаге является полностью правильным. Здесь же я могу давать советы и предложения.
</div>

<br/>

<div class="alert alert-warning">
    <h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> В случае, когда решение на отдельном шаге станет еще лучше, если внести небольшие коррективы.
</div> 
 

<br/>
<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера <a class="tocSkip"></h2>
 
   
<b>На доработку🤔:</b>
 В случае, когда решение на отдельном шаге требует существенной переработки и внесения правок. Напоминаю, что проект не может быть принят с первого раза, если ревью содержит комментарии, рекомендующие доработать шаги.
</div>  

<br/>
<div class="alert alert-info">
Рекомендую для твоих комментариев использовать синий или любой другой цветной фон. Я постараюсь ничего не пропустить, но если комментарии будут выделены, то это мне очень поможет!
</div>
    
Увидев у тебя неточность, в первый раз я лишь укажу на ее наличие и дам тебе возможность самостоятельно найти и исправить ее. На реальной работе твой руководитель будет поступать также. Но если ты пока не справишься с такой задачей - при следующей проверке я дам более точную подсказку.



# Проект "Секреты Темнолесья"

- Автор: Соловьев Иван Александрович
- Дата: 12.06.2025

### Цели и задачи проекта

**Цель проекта** - на основе предоставленных исторических данных, относящихся к игровой сфере, помочь коллегам изучить развитие этой индустрии.

Для этого следует решить следующие задачи:

1) Познакомиться с данными.

2) Проверить корректность данных, их тип, наличие пропусков и дубликатов.

3) Провести фильтрацию данных за период 2000-2013 гг.
    
4) Провести категоризацию данных.

### Описание данных

Данные /datasets/new_games.csv содержат информацию о продажах игр разных жанров и платформ, а также пользовательские и экспертные оценки игр:
Name — название игры.

**Platform** — название платформы.
    
**Year of Release** — год выпуска игры.

**Genre** — жанр игры.

**NA sales** — продажи в Северной Америке (в миллионах проданных копий).

**EU sales** — продажи в Европе (в миллионах проданных копий).

**JP sales** — продажи в Японии (в миллионах проданных копий).

**Other sales** — продажи в других странах (в миллионах проданных копий).

**Critic Score** — оценка критиков (от 0 до 100).

**User Score** — оценка пользователей (от 0 до 10).

**Rating** — рейтинг организации ESRB (англ. Entertainment Software Rating Board). Эта ассоциация определяет рейтинг компьютерных игр и присваивает им подходящую возрастную категорию.

### Содержимое проекта

- Загрузка и знакомство с данными
- Проверка ошибок в данных и их предобработка
- Фильтрация данных
- Категоризация данных
- Итоговый вывод

---

<div class="alert alert-success">
<h2> Комментарий ревьюера  <a class="tocSkip"> </h2>

👍 Отлично по описанию. Цели и задачи ясны. Контекст проекта будет понятен любому потенциальному читателю.
 



## 1. Загрузка данных и знакомство с ними

- Загрузим необходимые библиотеки Python и данные датасета `/datasets/new_games.csv`.


In [3]:
# Импортируем библиотеку pandas
import pandas as pd

In [4]:
# Выгружаем данные из датасета new_games.csv в датафрейм games
games = pd.read_csv('https://code.s3.yandex.net/datasets/new_games.csv')

- Изучим данные датасета, для этого выведем первые 5 строк и общую информацию о нем:


In [5]:
games.head()

Name Platform  Year of Release         Genre  NA sales  \
0                Wii Sports      Wii           2006.0        Sports     41.36   
1         Super Mario Bros.      NES           1985.0      Platform     29.08   
2            Mario Kart Wii      Wii           2008.0        Racing     15.68   
3         Wii Sports Resort      Wii           2009.0        Sports     15.61   
4  Pokemon Red/Pokemon Blue       GB           1996.0  Role-Playing     11.27   

  EU sales JP sales  Other sales  Critic Score User Score Rating  
0    28.96     3.77         8.45          76.0          8      E  
1     3.58     6.81         0.77           NaN        NaN    NaN  
2    12.76     3.79         3.29          82.0        8.3      E  
3    10.93     3.28         2.95          80.0          8      E  
4     8.89    10.22         1.00           NaN        NaN    NaN

In [6]:
games.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16956 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Name             16954 non-null  object 
 1   Platform         16956 non-null  object 
 2   Year of Release  16681 non-null  float64
 3   Genre            16954 non-null  object 
 4   NA sales         16956 non-null  float64
 5   EU sales         16956 non-null  object 
 6   JP sales         16956 non-null  object 
 7   Other sales      16956 non-null  float64
 8   Critic Score     8242 non-null   float64
 9   User Score       10152 non-null  object 
 10  Rating           10085 non-null  object 
dtypes: float64(4), object(7)
memory usage: 1.4+ MB


Выведем количество строк и столбцов, сохраним эти данные в переменную:

In [7]:
games_strcol = games.shape

**Промежуточный вывод:**

Датафрейм состоит из 16956 строк и 11 столбцов, то есть он является достаточно объемным.

В каждом столбце данные соответствуют своему описанию и логически не конфликтуют с названием колонки.

Наиболее частые пропуски наблюдаются в столбцах Critic Score, User Score, Rating: 8242, 10152 и 10085 пропущенных значений соответственно. Не столь критическое количество пропусков встречается в столбце Year of Release - 16681. По 2 пропущенных значения имеются в столбце Name и Genre.

Колонки с названиями Name, Platform, Genre, NA sales, Other sales, Critic Score, Rating имеют правильный тип данных: либо тип object для текстовой информации (например, наименование игры, платформы, рейтинг ESRB), либо тип float64 (например, для количества продаж и оценки критиков).

Проблема с типом данных наблюдается у оставшихся столбцов, это: Year of Release (логичен тип datetime64 для даты релиза, но можно привести к типу int, так как мы имеем только год и он не может обозначаться нецелым числом), EU sales и JP sales (логичен тип float64 для отображения количества продаж в миллионах), User Score (нужно привести к типу float64 для последующей обработки оценки пользователей).

*Дополнительные замечания:* рекомендуется привести названия столбцов к формату snake case для более удобного и привычного использования в дальнейшем. </font>

<div class="alert alert-success">
<h2> Комментарий ревьюера  <a class="tocSkip"> </h2>

👍 Все так. В данных есть пропуски, можно поправить по стилю названий колонок, а также обязательно нужно разобраться в проблемах с несоответствием типов данных содержанию столбцов.



---

## 2.  Проверка ошибок в данных и их предобработка


### 2.1. Названия, или метки, столбцов датафрейма

Выведем названия всех столбцов датафрейма:

In [8]:
games.columns

Index(['Name', 'Platform', 'Year of Release', 'Genre', 'NA sales', 'EU sales',
       'JP sales', 'Other sales', 'Critic Score', 'User Score', 'Rating'],
      dtype='object')

Приведем все столбцы к стилю snake case

In [9]:
games.columns = [x.lower().replace(' ', '_') for x in games.columns.values]

Выведем датафрейм с измененными названиями столбцов:

In [10]:
#df.columns
display(games)

name platform  year_of_release         genre  \
0                         Wii Sports      Wii           2006.0        Sports   
1                  Super Mario Bros.      NES           1985.0      Platform   
2                     Mario Kart Wii      Wii           2008.0        Racing   
3                  Wii Sports Resort      Wii           2009.0        Sports   
4           Pokemon Red/Pokemon Blue       GB           1996.0  Role-Playing   
...                              ...      ...              ...           ...   
16951  Samurai Warriors: Sanada Maru      PS3           2016.0        Action   
16952               LMA Manager 2007     X360           2006.0        Sports   
16953        Haitaka no Psychedelica      PSV           2016.0     Adventure   
16954               Spirits & Spells      GBA           2003.0      Platform   
16955            Winning Post 8 2016      PSV           2016.0    Simulation   

       na_sales eu_sales jp_sales  other_sales  critic_score user_score rating  
0         41.36    28.96     3.77         8.45          76.0          8      E  
1         29.08     3.58     6.81         0.77           NaN        NaN    NaN  
2         15.68    12.76     3.79         3.29          82.0        8.3      E  
3         15.61    10.93     3.28         2.95          80.0          8      E  
4         11.27     8.89    10.22         1.00           NaN        NaN    NaN  
...         ...      ...      ...          ...           ...        ...    ...  
16951      0.00      0.0     0.01         0.00           NaN        NaN    NaN  
16952      0.00     0.01      0.0         0.00           NaN        NaN    NaN  
16953      0.00      0.0     0.01         0.00           NaN        NaN    NaN  
16954      0.01      0.0      0.0         0.00           NaN        NaN    NaN  
16955      0.00      0.0     0.01         0.00           NaN        NaN    NaN  

[16956 rows x 11 columns]

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

👍 Верно и оптимально по переименованию.  Отлично, что делаешь проверку результата преобразования.



**Промежуточный вывод:**

На данном этапе для дальнейшей удобной работы с данными датафрейма названия столбцов преобразованы в стиль написания snake case.

### 2.2. Типы данных

- Если встречаются некорректные типы данных, предположите их причины.
- При необходимости проведите преобразование типов данных. Помните, что столбцы с числовыми данными и пропусками нельзя преобразовать к типу `int64`. Сначала вам понадобится обработать пропуски, а затем преобразовать типы данных.

Некорректные типы данных могут быть вызваны различными причинами.

Например, можно предположить, что столбец year_of_release заполнялся неопытным человеком, из-за пропусков в числовых данных столбец автоматически приобрел тип `float64` (так как NaN относится к этому типу) или это особенности определенного метода (dt.year), с помощью которого формировался этот столбец, поэтому от типа `datetime64` данные преобразовались во `float64`. В столбцах с продажами `eu_sales` и `jp_sales` тип данных определен как *object*, возможно, отсутствующие данные обозначались строкой *unknown* или чем-то подобным, поэтому получился столбец со смешанными типами данных, отчего он определился как *object*. Аналогичная ситуация и с данными в столбце `user_score`.

Попробуем привести данные к нужным типам, пользуясь методом to_numeric:

(параметр `errors='coerce'` используется для того, чтобы в случае наличия в столбцах строковых данных, они преобразовались в пропуски NaN)

In [11]:
for column in ['year_of_release', 'eu_sales', 'jp_sales', 'user_score']:
    games[column] = pd.to_numeric(games[column], errors='coerce')

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

👍 Отлично. Также можно использовать apply, чтобы написать все в одну строку:
    
`df[['eu_sales', 'jp_sales','user_score']] = df[['eu_sales', 'jp_sales','user_score']].apply(pd.to_numeric, errors='coerce')`



Проверим типы данных обновленных столбцов:

In [12]:
games.dtypes

name                object
platform            object
year_of_release    float64
genre               object
na_sales           float64
eu_sales           float64
jp_sales           float64
other_sales        float64
critic_score       float64
user_score         float64
rating              object
dtype: object

Выведем первые 5 строк датафрейма:

In [13]:
games.head()

name platform  year_of_release         genre  na_sales  \
0                Wii Sports      Wii           2006.0        Sports     41.36   
1         Super Mario Bros.      NES           1985.0      Platform     29.08   
2            Mario Kart Wii      Wii           2008.0        Racing     15.68   
3         Wii Sports Resort      Wii           2009.0        Sports     15.61   
4  Pokemon Red/Pokemon Blue       GB           1996.0  Role-Playing     11.27   

   eu_sales  jp_sales  other_sales  critic_score  user_score rating  
0     28.96      3.77         8.45          76.0         8.0      E  
1      3.58      6.81         0.77           NaN         NaN    NaN  
2     12.76      3.79         3.29          82.0         8.3      E  
3     10.93      3.28         2.95          80.0         8.0      E  
4      8.89     10.22         1.00           NaN         NaN    NaN

**Промежуточный вывод:**

На данном этапе отмечены возможные причины неправильного определения типов данных, осуществлено их успешное промежуточное преобразование, при этом возможные встречи текстовых данных в столбцах с числами ликвидированы.

Дополнительные замечания: требуется более детальная работа с пропусками, их возможное удаление или минимизация.

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

👍 Все верно.
    
Замечательно по оформлению и по аргументации.

### 2.3. Наличие пропусков в данных

Посчитаем количество пропусков в каждом столбце в абсолютных и относительных значениях.


In [14]:
games.isna().sum()

name                  2
platform              0
year_of_release     275
genre                 2
na_sales              0
eu_sales              6
jp_sales              4
other_sales           0
critic_score       8714
user_score         9268
rating             6871
dtype: int64

In [15]:
games.isna().sum() / games.shape[0]

name               0.000118
platform           0.000000
year_of_release    0.016218
genre              0.000118
na_sales           0.000000
eu_sales           0.000354
jp_sales           0.000236
other_sales        0.000000
critic_score       0.513918
user_score         0.546591
rating             0.405225
dtype: float64

Наибольшее число пропусков характерно для столбцов `user_score` (54,7%), `critic_score` (51,4% от всех данных), `rating` (40,5%). Совсем мало пропусков имеет столбец `year_of_release` - всего 1,6% от всего объема датафрейма, а также от 2 до 6 пропусков (из 16956 строк датафрейма) наблюдаются в `name`, `genre` и `eu_sales`, `jp_sales`.

Можно предположить, что в данных представлена информация об играх за период, когда рейтинги от пользователей или критиков не собирались, поэтому имеется много пропусков в колонках, относящихся к этим данным. Пропуски в столбце `rating` можно объяснить тем, что организация ESRB начала выдавать рейтинги играм только в 1994 году, а, как уже было сказано, в датафрейме имеются даннные за периоды, предшествовавшие этому. Некоторые игры могли разрабатываться, но в итоге так и не быть представлеными публике - так, предположительно, описывается незначительно число незаполненных данных в столбеце с датой релиза.

Можно проверить наиболее ранний год выпуска игры в имеющихся данных, чтобы быть уверенным в своих предположениях:

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

👍 Согласен с тобой. Думаю, здесь может быть гипотеза в том, что для слишком старых игр или для игр из других регионов, не переведенных на английский язык, не существует достаточного количества оценок критиков и пользователей по естественным причинам. 



- Изучите данные с пропущенными значениями. Напишите промежуточный вывод: для каких столбцов характерны пропуски и сколько их. Предположите, почему пропуски могли возникнуть. Укажите, какие действия с этими данными можно сделать и почему.


In [16]:
games['eu_sales'].unique()

array([2.896e+01, 3.580e+00, 1.276e+01, 1.093e+01, 8.890e+00, 2.260e+00,
       9.140e+00, 9.180e+00, 6.940e+00, 6.300e-01, 1.095e+01, 7.470e+00,
       6.180e+00, 8.030e+00, 4.890e+00, 8.490e+00, 9.090e+00, 4.000e-01,
       3.750e+00, 9.200e+00, 4.460e+00, 2.710e+00, 3.440e+00, 5.140e+00,
       5.490e+00, 3.900e+00, 5.350e+00, 3.170e+00, 5.090e+00, 4.240e+00,
       5.040e+00, 5.860e+00, 3.680e+00, 4.190e+00, 5.730e+00, 3.590e+00,
       4.510e+00, 2.550e+00, 4.020e+00, 4.370e+00, 6.310e+00, 3.450e+00,
       2.810e+00, 2.850e+00, 3.490e+00, 1.000e-02, 3.350e+00, 2.040e+00,
       3.070e+00, 3.870e+00, 3.000e+00, 4.820e+00, 3.640e+00, 2.150e+00,
       3.690e+00, 2.650e+00, 2.560e+00, 3.110e+00, 3.140e+00, 1.940e+00,
       1.950e+00, 2.470e+00, 2.280e+00, 3.420e+00, 3.630e+00, 2.360e+00,
       1.710e+00, 1.850e+00, 2.790e+00, 1.240e+00, 6.120e+00, 1.530e+00,
       3.470e+00, 2.240e+00, 5.010e+00, 2.010e+00, 1.720e+00, 2.070e+00,
       6.420e+00, 3.860e+00, 4.500e-01, 3.480e+00, 

<div class="alert alert-warning">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> 
    
Количественные признаки лучше не выводить сплошняком на принт. Оно бесполезно для анализа и больно для глаз.
    
Такие ряды чисел называются простынями и являются визуальным мусором. В финальном проекте лучше его не оставлять. В дальнейших проектах для анализа количественных признаков будем строить гистограммы.

In [17]:
games['jp_sales'].unique()

array([3.770e+00, 6.810e+00, 3.790e+00, 3.280e+00, 1.022e+01, 4.220e+00,
       6.500e+00, 2.930e+00, 4.700e+00, 2.800e-01, 1.930e+00, 4.130e+00,
       7.200e+00, 3.600e+00, 2.400e-01, 2.530e+00, 9.800e-01, 4.100e-01,
       3.540e+00, 4.160e+00, 6.040e+00, 4.180e+00, 3.840e+00, 6.000e-02,
       4.700e-01, 5.380e+00, 5.320e+00, 5.650e+00, 1.870e+00, 1.300e-01,
       3.120e+00, 3.600e-01, 1.100e-01, 4.350e+00, 6.500e-01, 7.000e-02,
       8.000e-02, 4.900e-01, 3.000e-01, 2.660e+00, 2.690e+00, 4.800e-01,
       3.800e-01, 5.330e+00, 1.910e+00, 3.960e+00, 3.100e+00, 1.100e+00,
       1.200e+00, 1.400e-01, 2.540e+00, 2.140e+00, 8.100e-01, 2.120e+00,
       4.400e-01, 3.150e+00, 1.250e+00, 4.000e-02, 0.000e+00, 2.470e+00,
       2.230e+00, 1.690e+00, 1.000e-02, 3.000e+00, 2.000e-02, 4.390e+00,
       1.980e+00, 1.000e-01, 3.810e+00, 5.000e-02, 2.490e+00, 1.580e+00,
       3.140e+00, 2.730e+00, 6.600e-01, 2.200e-01, 3.630e+00, 1.450e+00,
       1.310e+00, 2.430e+00, 7.000e-01, 3.500e-01, 

Вывод состоит в том, что в столбцах с продажами по Европе и Японии не содержится аномальных и отрицательных значений, а также строк. Можно переходить к обработке пропусков в этих столбцах.

- Обработайте пропущенные значения. Для каждого случая вы можете выбрать оптимальный, на ваш взгляд, вариант: заменить на определённое значение, оставить как есть или удалить.
- Если вы решите заменить пропуски на значение-индикатор, то убедитесь, что предложенное значение не может быть использовано в данных.
- Если вы нашли пропуски в данных с количеством проданных копий игры в том или ином регионе, их можно заменить на среднее значение в зависимости от названия платформы и года выхода игры.

In [18]:
games['year_of_release'].min()

1980.0

Пропуски в столбцах `name`, `genre` можно удалить, так как их количество относительно всего датафрейма очень мало, менее 1%, это не повлияет на дальнейшую обработку данных.

In [19]:
games = games.dropna(subset=['name'])
games = games.dropna(subset=['genre'])

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

👍 

Пропуски в столбцах `eu_sales` и `jp_sales` заменим на среднее значение в зависимости от названия платформы и года выхода игры.

In [20]:
mean_sales = games.groupby(['year_of_release', 'platform'])['eu_sales'].transform('mean')
games['eu_sales'] = games['eu_sales'].fillna(mean_sales)

mean_sales_jp = games.groupby(['year_of_release', 'platform'])['jp_sales'].transform('mean')
games['jp_sales'] = games['jp_sales'].fillna(mean_sales_jp)

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

👍 Оптимальное решение.

Пропуски в столбце с датой релиза `year_of_release` оставляем пропуски в их изначальном варианте.

Пропуски в столбцах `user_score` и `critic_score` заменим на -1, что будет означать отсутствие пользовательских или критических отзывов.

In [21]:
games['user_score'] = games['user_score'].fillna(-1)
games['critic_score'] = games['critic_score'].fillna(-1)

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

👍 

Пропуски в столбце с рейтингом ESRB `rating` можно заменить на пустую строку, так как в столбце содержится строковый тип данных.

In [22]:
games['rating'] = games['rating'].fillna('')

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

👍 В строковом типе у нас есть возможность заполнить пропуски строкой с чем-то осмысленным и человекочитаемым. Обычно пишут «неизвестно» или «unknown» или просто «unk».

Проверим наличие пропусков в обновленном датафрейме:

In [23]:
games.isna().sum()

name                 0
platform             0
year_of_release    275
genre                0
na_sales             0
eu_sales             0
jp_sales             0
other_sales          0
critic_score         0
user_score           0
rating               0
dtype: int64

**Промежуточный вывод:**

На данном этапе проведена обработка пропусков, отсутствующие значения заменились на подходящие по логике индикаторы.

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

👍 В целом можем остаивть пропуски в годе как есть, а можно и удалить.
    
Тогда тоже можно все дропнуть одной строкой `df = df.dropna(subset=[‘year_of_Release','name', 'genre'])`

### 2.4. Явные и неявные дубликаты в данных

- Изучите уникальные значения в категориальных данных, например с названиями жанра игры, платформы, рейтинга и года выпуска. Проверьте, встречаются ли среди данных неявные дубликаты, связанные с опечатками или разным способом написания.
- При необходимости проведите нормализацию данных с текстовыми значениями. Названия или жанры игр можно привести к нижнему регистру, а названия рейтинга — к верхнему.

In [24]:
print('Список лет выпуска игр:\n', games['year_of_release'].sort_values().unique())
print('Количество уникальных позиций:', games['year_of_release'].nunique())

Список лет выпуска игр:
 [1980. 1981. 1982. 1983. 1984. 1985. 1986. 1987. 1988. 1989. 1990. 1991.
 1992. 1993. 1994. 1995. 1996. 1997. 1998. 1999. 2000. 2001. 2002. 2003.
 2004. 2005. 2006. 2007. 2008. 2009. 2010. 2011. 2012. 2013. 2014. 2015.
 2016.   nan]
Количество уникальных позиций: 37


In [25]:
print('Список игровых платформ:\n', games['platform'].unique())
print('Количество уникальных позиций:', games['platform'].nunique())

Список игровых платформ:
 ['Wii' 'NES' 'GB' 'DS' 'X360' 'PS3' 'PS2' 'SNES' 'GBA' 'PS4' '3DS' 'N64'
 'PS' 'XB' 'PC' '2600' 'PSP' 'XOne' 'WiiU' 'GC' 'GEN' 'DC' 'PSV' 'SAT'
 'SCD' 'WS' 'NG' 'TG16' '3DO' 'GG' 'PCFX']
Количество уникальных позиций: 31


In [26]:
print('Список жанров:\n', games['genre'].unique())
print('Количество уникальных позиций:', games['genre'].nunique())

Список жанров:
 ['Sports' 'Platform' 'Racing' 'Role-Playing' 'Puzzle' 'Misc' 'Shooter'
 'Simulation' 'Action' 'Fighting' 'Adventure' 'Strategy' 'MISC'
 'ROLE-PLAYING' 'RACING' 'ACTION' 'SHOOTER' 'FIGHTING' 'SPORTS' 'PLATFORM'
 'ADVENTURE' 'SIMULATION' 'PUZZLE' 'STRATEGY']
Количество уникальных позиций: 24


Сохраним количество уникальных позиций до предобработки дубликатов:

In [27]:
genre_prev_cnt = games['genre'].nunique()

In [28]:
print('Список рейтинга ESRB:\n', games['rating'].unique())
print('Количество уникальных позиций:', games['rating'].nunique())

Список рейтинга ESRB:
 ['E' '' 'M' 'T' 'E10+' 'K-A' 'AO' 'EC' 'RP']
Количество уникальных позиций: 9


 <div class="alert alert-warning">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> 

Здесь хитрая ситуация с rating. У нас есть игры как из Европы, так и из Азии. В основном используется рейтинг ESRB https://ru.wikipedia.org/wiki/Entertainment_Software_Rating_Board 
    
Но некоторых категорий из представленных у нас в нем нет. Фактически, это сходные категории, но в другом обозначении.
    
Например, можно заменить значения 'К-А' на 'Е'.
    
Рекомендую в подобных случаях уточнять информацию в интернете (в реальных проектах на реальной работе также, ну и можно всегда спрашивать у коллег, особенно у тех, кто предоставил данные. Часто такие категории объясняются в паспорте данных. Ну в нашем случае нужно залезть в интернет). 



*Замечание:* в столбце `genre` имеются неявные дубликаты: один и те же позиции дублируются в нижнем и верхнем регистрах, поэтому требуется провести их нормализацию. Приведем данные в этом столбце к нижнему регистру:

In [29]:
games['genre'] = games['genre'].str.lower()

После изменения регистра выведем новый список уникальных значений и их количество:

In [30]:
print('Список жанров:\n', games['genre'].unique())
print('Количество уникальных позиций:', games['genre'].nunique())

Список жанров:
 ['sports' 'platform' 'racing' 'role-playing' 'puzzle' 'misc' 'shooter'
 'simulation' 'action' 'fighting' 'adventure' 'strategy']
Количество уникальных позиций: 12


Сохраним количество уникальных позиций после нормализации данных:

In [31]:
genre_new_cnt = games['genre'].nunique()

Видим, что количество уникальных значений снизилось в два раза. Теперь данные в этом столбце нормализованы.

Перейдем к следующему шагу: проверим наличие явных дубликатов в данных. Для этого используется метод duplicated:

In [32]:
games[games.duplicated()]

name platform  \
268                            Batman: Arkham Asylum      PS3   
368                 James Bond 007: Agent Under Fire      PS2   
717                            God of War: Ascension      PS3   
823                                Wipeout: The Game      Wii   
848                  Rayman Raving Rabbids: TV Party      Wii   
...                                              ...      ...   
16671        Fullmetal Alchemist: Prince of the Dawn      Wii   
16753                                      Routes PE      PS2   
16799                            Transformers: Prime      Wii   
16912  Metal Gear Solid V: The Definitive Experience     XOne   
16940                          The Longest 5 Minutes      PSV   

       year_of_release      genre  na_sales  eu_sales  jp_sales  other_sales  \
268             2009.0     action      2.24      1.31      0.07         0.61   
368             2001.0    shooter      1.90      1.13      0.10         0.41   
717             2013.0     action      1.23      0.63      0.04         0.35   
823             2009.0       misc      1.94      0.00      0.00         0.12   
848             2008.0       misc      0.72      1.08      0.00         0.23   
...                ...        ...       ...       ...       ...          ...   
16671           2009.0  adventure      0.00      0.00      0.01         0.00   
16753           2007.0  adventure      0.00      0.00      0.01         0.00   
16799           2012.0     action      0.00      0.01      0.00         0.00   
16912           2016.0     action      0.01      0.00      0.00         0.00   
16940           2016.0     action      0.00      0.00      0.01         0.00   

       critic_score  user_score rating  
268            91.0         8.9      T  
368            72.0         7.9      T  
717            80.0         7.5      M  
823            -1.0        -1.0         
848            73.0         7.7   E10+  
...             ...         ...    ...  
16671          -1.0        -1.0         
16753          -1.0        -1.0         
16799          -1.0        -1.0         
16912          -1.0        -1.0      M  
16940          -1.0        -1.0         

[241 rows x 11 columns]

 <div class="alert alert-warning">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> 

также можно не выводить всю таблицу. Оно нам не помогает. Достаточно просто вывести одну чиселку с количеством дубликатов - 241.

Удостоверимся на одном примере в том, что имеем дело с явными дубликатами:

In [33]:
games[games['name']=='Batman: Arkham Asylum']

name platform  year_of_release   genre  na_sales  \
267   Batman: Arkham Asylum      PS3           2009.0  action      2.24   
268   Batman: Arkham Asylum      PS3           2009.0  action      2.24   
381   Batman: Arkham Asylum     X360           2009.0  action      2.20   
5666  Batman: Arkham Asylum       PC           2009.0  action      0.00   

      eu_sales  jp_sales  other_sales  critic_score  user_score rating  
267       1.31      0.07         0.61          91.0         8.9      T  
268       1.31      0.07         0.61          91.0         8.9      T  
381       0.95      0.02         0.31          92.0         8.7      T  
5666      0.28      0.00         0.05          91.0         8.7      T

Удалим явные дубликаты и сохраним новый датафрейм:

In [34]:
drop_dup_games = games.drop_duplicates()
drop_dup_games.shape

(16713, 11)

**Выведем все изменения, произошедшие с датафреймом при работе с пропусками и дубликатами:**

Предобработка неявных дубликатов, а именно данных в столбце `genre`, повлияла на датафрейм следующим образом:

In [40]:
deleted_str = games.shape[0] - drop_dup_games.shape[0]
share_deleted_str = deleted_str / games.shape[0]

In [41]:
print('Количество удаленных строк в абсолютном значении:', deleted_str)
print('Количество удаленных строк в относительном значении:', share_deleted_str)

Количество удаленных строк в абсолютном значении: 241
Количество удаленных строк в относительном значении: 0.014214934528724785


<div class="alert alert-info">
<h2> Комментарий студента <a class="tocSkip"></h2>

Замечение поправлено

<div class="alert alert-success">
<h2> Комментарий ревьюера #2 <a class="tocSkip"> </h2>

👍 Cнова привет. Спасибо за заметное оформление твоих комментариев.
    
Все верно по расчету.
</div>


 <div class="alert alert-block alert-danger">
<h2><s> Комментарий ревьюера <a class="tocSkip"></h2>
  
<b>На доработку🤔:</b>

`Количество удаленных строк в относительном значении: 0.5` - это 50%? Кажется, не соответствует действительности.
    
***
    
Смотри ,на этом этапе мы решаем задачу узнать, а не много ли строк мы порезали за всю предобработку. Достаточно на моменте загрузки датасета замерить сколько было строк, а здесь после всей-всей предобработки замерить, сколько осталось. Нужно посмотреть разность и долю.

Предобработка пропусков и явных дубликатов в общем повлияла на датафрейм следующим образом:

In [39]:
print('Количество удаленных строк в абсолютном значении:', games_strcol[0] - drop_dup_games.shape[0])
print('Количество удаленных строк в относительном значении:', round((games.shape[0] - drop_dup_games.shape[0])/games.shape[0], 3))

Количество удаленных строк в абсолютном значении: 243
Количество удаленных строк в относительном значении: 0.014


Ищем неявные дубликаты через метод .duplicated. Осуществим поиск для столбцов с наименованием игры и платформой:

In [103]:
print(drop_dup_games[drop_dup_games.duplicated(subset=['name', 'platform'])])
print('Количество неявных дубликатов по столбцам [name] + [platform]:', drop_dup_games[drop_dup_games.duplicated(subset=['name', 'platform'])].shape[0])

                              name platform  year_of_release     genre  \
1605   Need for Speed: Most Wanted     X360           2005.0    racing   
4173            Sonic the Hedgehog      PS3              NaN  platform   
11873  Need for Speed: Most Wanted       PC           2012.0    racing   
16465                Madden NFL 13      PS3           2012.0    sports   

       na_sales  eu_sales  jp_sales  other_sales  critic_score  user_score  \
1605        1.0      0.13      0.02         0.10          83.0         8.5   
4173        0.0      0.48      0.00         0.00          43.0         4.1   
11873       0.0      0.06      0.00         0.02          82.0         8.5   
16465       0.0      0.01      0.00         0.00          83.0         5.5   

      rating  
1605       T  
4173    E10+  
11873      T  
16465      E  
Количество неявных дубликатов по столбцам [name] + [platform]: 4


Видим, что по указанным столбцам имеются 4 дубликата, удалим их:

In [104]:
drop_dup_games = drop_dup_games.drop_duplicates(subset=['name', 'platform'])

Проверим, что удаление произошло успешно:

In [105]:
print('Количество неявных дубликатов по столбцам [name] + [platform]:', drop_dup_games[drop_dup_games.duplicated(subset=['name', 'platform'])].shape[0])

Количество неявных дубликатов по столбцам [name] + [platform]: 0


Осуществим аналогичный поиск по имени, платформе и году релиза игры:

In [106]:
print(drop_dup_games[drop_dup_games.duplicated(subset=['name', 'platform', 'year_of_release'])])
print('Количество неявных дубликатов по столбцам [name] + [platform]:', drop_dup_games[drop_dup_games.duplicated(subset=['name', 'platform', 'year_of_release'])].shape[0])

Empty DataFrame
Columns: [name, platform, year_of_release, genre, na_sales, eu_sales, jp_sales, other_sales, critic_score, user_score, rating]
Index: []
Количество неявных дубликатов по столбцам [name] + [platform]: 0


Неявных дубликатов в указанных трех столбцах не обнаружено, что говорит о том, что вероятнее всего больше неявных дубликатов в датафрейме не имеется.

**Промежуточный вывод:**

В данных выявились как неявные дубликаты, так и явные дубликаты. Неявные дубликаты сосредоточились в колонке genre, они обработаны и приведены к унифицированной форме. В остальных столбцах неявных дубликатов не обнаружилось. Явные дубликаты исследованы с помощью метода библиотеки pandas, удалены все повторения одной и той же строки. Обработанный датафрейм сохранен в новый `drop_dup_games`, имеющий после манипуляций 16713 строк (по сравнению с 16956 строк в старом датафрейме).

**Общий вывод по предобработке данных:**

Данные были загружены из предоставленного файла .csv, было проведено знакомство с ними и их типами данных. Столбцы были преобразованы в предпочтительный для работы с кодом стиль snake case. Проведена работа с преобразованием типов данных в логически верные. Пропуски были исследованы и в зависимости от условий удалены, оставлены или заполнены значением-индикатором. Также были обнаружены как неявные, так и явные дубликаты, которые были устранены.

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

👍 А по оформлению замечательно.

---

## 3. Фильтрация данных

Коллеги хотят изучить историю продаж игр в начале XXI века, и их интересует период с 2000 по 2013 год включительно. Отберите данные по этому показателю. Сохраните новый срез данных в отдельном датафрейме, например `df_actual`.

Так как нас интересуют данные только за период с 2000 по 2013 год включительно, сделаем срез нужных нам данных:

In [116]:
df_actual = drop_dup_games[(drop_dup_games['year_of_release'] >= 2000) & (drop_dup_games['year_of_release'] <= 2013)]

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

👍

In [109]:
df_actual.head(10)

name platform  year_of_release       genre  na_sales  \
0                  Wii Sports      Wii           2006.0      sports     41.36   
2              Mario Kart Wii      Wii           2008.0      racing     15.68   
3           Wii Sports Resort      Wii           2009.0      sports     15.61   
6       New Super Mario Bros.       DS           2006.0    platform     11.28   
7                    Wii Play      Wii           2006.0        misc     13.96   
8   New Super Mario Bros. Wii      Wii           2009.0    platform     14.44   
10                 Nintendogs       DS           2005.0  simulation      9.05   
11              Mario Kart DS       DS           2005.0      racing      9.71   
13                    Wii Fit      Wii           2007.0      sports      8.92   
14         Kinect Adventures!     X360           2010.0        misc     15.00   

    eu_sales  jp_sales  other_sales  critic_score  user_score rating  
0      28.96      3.77         8.45          76.0         8.0      E  
2      12.76      3.79         3.29          82.0         8.3      E  
3      10.93      3.28         2.95          80.0         8.0      E  
6       9.14      6.50         2.88          89.0         8.5      E  
7       9.18      2.93         2.84          58.0         6.6      E  
8       6.94      4.70         2.24          87.0         8.4      E  
10     10.95      1.93         2.74          -1.0        -1.0         
11      7.47      4.13         1.90          91.0         8.6      E  
13      8.03      3.60         2.15          80.0         7.7      E  
14      4.89      0.24         1.69          61.0         6.3      E

И выведем его размеры:

In [111]:
df_actual.shape

(12778, 11)

Для того чтобы быть более уверенным в новом датафрейме, попробуем найти в нем данные за года, не включенные в промежуток:

In [113]:
df_actual[(df_actual['year_of_release'] < 2000) | (df_actual['year_of_release'] > 2013)]

Empty DataFrame
Columns: [name, platform, year_of_release, genre, na_sales, eu_sales, jp_sales, other_sales, critic_score, user_score, rating]
Index: []

Таких данных нет, а значит датафрейм сформирован верно.

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

👍

**Промежуточный вывод:**

На данном этапе сформирован новый датафрейм путем фильтрации предобработанного исходного.

---

## 4. Категоризация данных
    
Проведите категоризацию данных:
- Разделите все игры по оценкам пользователей и выделите такие категории: высокая оценка (от 8 до 10 включительно), средняя оценка (от 3 до 8, не включая правую границу интервала) и низкая оценка (от 0 до 3, не включая правую границу интервала).
- Разделите все игры по оценкам критиков и выделите такие категории: высокая оценка (от 80 до 100 включительно), средняя оценка (от 30 до 80, не включая правую границу интервала) и низкая оценка (от 0 до 30, не включая правую границу интервала).
- Выделите топ-7 платформ по количеству игр, выпущенных за весь актуальный период.

- Разделите все игры по оценкам критиков и выделите такие категории: высокая оценка (от 80 до 100 включительно), средняя оценка (от 30 до 80, не включая правую границу интервала) и низкая оценка (от 0 до 30, не включая правую границу интервала).

Для разделения всех игр по оценкам пользователей и выделения нужных категорий напишем пользовательскую функцию и применим ее для рассматриваемого столбца `user_score:`

In [119]:
# напишем функцию для разделения оценок пользователей по категориям
def category_user(value):
    if 0 <= value < 3:
        return "Низкая оценка"
    elif 3 <= value < 8:
        return "Средняя оценка"
    elif 8 <= value <= 10:
        return "Высокая оценка"
    else:
        return "Нет оценки"

df_actual = df_actual.copy()
df_actual['user_category'] = df_actual['user_score'].apply(category_user)

Для данных в столбце `critic_score` выберем другой способ категоризации и применим метод pd.cut:

In [121]:
df_actual['critic_category'] = pd.cut(df_actual['critic_score'], bins=[-1, 0, 30, 80, 100.1],\
                                             labels=["Нет оценки", "Низкая оценка", "Средняя оценка", "Высокая оценка"], right=False)

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

👍 Оба подхода верные.

In [122]:
df_actual.head(10)

name platform  year_of_release       genre  na_sales  \
0                  Wii Sports      Wii           2006.0      sports     41.36   
2              Mario Kart Wii      Wii           2008.0      racing     15.68   
3           Wii Sports Resort      Wii           2009.0      sports     15.61   
6       New Super Mario Bros.       DS           2006.0    platform     11.28   
7                    Wii Play      Wii           2006.0        misc     13.96   
8   New Super Mario Bros. Wii      Wii           2009.0    platform     14.44   
10                 Nintendogs       DS           2005.0  simulation      9.05   
11              Mario Kart DS       DS           2005.0      racing      9.71   
13                    Wii Fit      Wii           2007.0      sports      8.92   
14         Kinect Adventures!     X360           2010.0        misc     15.00   

    eu_sales  jp_sales  other_sales  critic_score  user_score rating  \
0      28.96      3.77         8.45          76.0         8.0      E   
2      12.76      3.79         3.29          82.0         8.3      E   
3      10.93      3.28         2.95          80.0         8.0      E   
6       9.14      6.50         2.88          89.0         8.5      E   
7       9.18      2.93         2.84          58.0         6.6      E   
8       6.94      4.70         2.24          87.0         8.4      E   
10     10.95      1.93         2.74          -1.0        -1.0          
11      7.47      4.13         1.90          91.0         8.6      E   
13      8.03      3.60         2.15          80.0         7.7      E   
14      4.89      0.24         1.69          61.0         6.3      E   

     user_category critic_category  
0   Высокая оценка  Средняя оценка  
2   Высокая оценка  Высокая оценка  
3   Высокая оценка  Высокая оценка  
6   Высокая оценка  Высокая оценка  
7   Средняя оценка  Средняя оценка  
8   Высокая оценка  Высокая оценка  
10      Нет оценки      Нет оценки  
11  Высокая оценка  Высокая оценка  
13  Средняя оценка  Высокая оценка  
14  Средняя оценка  Средняя оценка

In [123]:
df_actual.groupby('user_category')['user_score'].count().sort_values(ascending=False)

user_category
Нет оценки        6298
Средняя оценка    4080
Высокая оценка    2284
Низкая оценка      116
Name: user_score, dtype: int64

In [124]:
df_actual.groupby('critic_category', observed=True)['critic_score'].count().sort_values(ascending=False)

critic_category
Нет оценки        5612
Средняя оценка    5422
Высокая оценка    1689
Низкая оценка       55
Name: critic_score, dtype: int64

Выведем топ-7 платформ с наибольшим числом игр, выпущенных за рассматриваемый период:

In [125]:
df_top7 = df_actual.groupby('platform', as_index=False)['name'].count()
df_top7.sort_values(by='name', ascending=False).iloc[:7]

platform  name
9       PS2  2127
2        DS  2120
15      Wii  1275
12      PSP  1180
17     X360  1120
10      PS3  1086
4       GBA   811

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

👍 Все отлично.

---

## 5. Итоговый вывод

В конце напишите основной вывод и отразите, какую работу проделали. Не забудьте указать описание среза данных и новых полей, которые добавили в исходный датасет.

**Промежуточный вывод:**

Данные в датафрейме `df_actual` разделены на категории в зависимости от оценки пользователей и критиков. Также представлен топ-7 платформ по количеству игр за актуальный период 2000-2013 годов.

В проекте последовательно проведены знакомство с данными и их типом, изменение типов данных, предобработка пропусков и дубликатов, фильтрация и категоризация данных. В итоге из 16956 строк и 11 столбцов исходный датафрейм преобразовался в новый с 12778 строками и 13 столбцами. Такие изменения обусловлены:

1. Очисткой датафрейма от пропусков и дубликатов, в результате чего строк стало меньше.
2. Фильтрацией данных, в результате чего количество строк также уменьшилось.
3. Выделением категорий по столбцам `user_score` (пользовательские оценки) и `critic_score` (оценки критиков), в результате чего количество столбцов увеличилось.

In [127]:
display(games_strcol)
display(df_actual.shape)

(16956, 11)

(12778, 13)

На выходе получен датафрейм, с помощью которого коллегам будет легче изучить развитие игровой индустрии, опираясь на продажи игр в разных регионах, их жанр, оценки пользователей и критиков и соответствующие им категории.

Датафрейм очищен от ненужных данных и приведен к предпочтительной для анализа форме.

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>
 
👍 Проект о предобработке, в общем выводе описываем, что делаем по предобработке. Все логично, отлично по оформлению. 
</div>    



<div style="border:solid Chocolate 2px; padding: 40px">

**Общий вывод по проекту**
    
Спасибо за твой проект! Внимательная работа, практически все решения верные и оптимальные, но кое-что нужно доработать и кое-что можно улучшить, чтобы проект стал совсем замечательным.
    
**Отмечу положительные моменты проекта🚀**
    
* Отличное оформление. Есть описание, комментарии, выводы.
* Хорошее владение кодом, инструментами предобработки. Все сложные моменты выполнены отлично.
* Хорошая аргументация решений. Во всех случаях понятно, почему принимали те или иные решения.
    
**На что стоит обратить внимание🤔**
   
* Необходимо убрать лишнее и некорректное в шаге с неявными дубликатами. Там достаточно нормализовать данные, уточнить по идентичным категориям. Затем считаем дубликаты, удаляем дубликаты, считаем количество и долю удаленных строк за всю предобработку. 
* Эти и остальные замечания я также написал по ходу проекта.
    
**Желаю удачи и жду твой проект на повторное ревью!**😉
    
    
</div>

<div style="border:solid Chocolate 2px; padding: 40px">

**Комментарий ко второй проверке**

Спасибо за доработку. 🤝

Критичных комментариев нет. 
    
Если что-то пропустил или остались вопросы, напиши в телеграм @kriill

Желаю удачи в дальнейшей учебе!

</div>
